### PR TITLE
[SPARK-29464][PYTHON][ML] PySpark ML should expose Params.clear() to unset a user supplied Param

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
change PySpark ml ```Params._clear``` to ```Params.clear```

### Why are the changes needed?
PySpark ML currently has a private _clear() method that will unset a param. This should be made public to match the Scala API and give users a way to unset a user supplied param. 

### Does this PR introduce any user-facing change?
Yes. PySpark ml ```Params._clear``` ---> ```Params.clear```

### How was this patch tested?
Add test.

### CODE REVIEW

**Comment by huaxingao** [comment]
@BryanCutler @srowen 
Could you please review? Thanks a lot!

**Comment by BryanCutler** [comment]
@huaxingao I think there is a problem when the Java object has the param set, then it is cleared in Python but not Java. For example:

```python
In [4]:     >>> from pyspark.ml.linalg import Vectors 
   ...:     >>> df = spark.createDataFrame([(Vectors.dense([1.0]),), (Vectors.dense([2.0]),)], ["a"])
   ...:     >>> maScaler = MaxAbsScaler(inputCol="a", outputCol="scaled") 
   ...:     >>> model = maScaler.fit(df) 
   ...:     >>> model.setOutputCol("scaledOutput")                                                  
Out[4]: MaxAbsScaler_f91118f1dd81                                               

In [7]: model.clear(model.outputCol)                                                               

In [9]: model.getOutputCol()                                                                        
Out[9]: 'MaxAbsScaler_f91118f1dd81__output'

In [11]: model.transform(df).show()                                                                 
+-----+------+
|    a|scaled|
+-----+------+
|[1.0]| [0.5]|
|[2.0]| [1.0]|
+-----+------+
```

**Review by BryanCutler — CHANGES_REQUESTED** [review]
Need to sync with Java after clear

**Review by srowen — APPROVED** [review]
Looks OK to me pending tests if it looks good to you both.

**Comment by zhengruifeng** [comment]
retest this please

File: python/pyspark/ml/tests/test_param.py [hunk 1]
```diff
@@ -248,6 +252,18 @@ def test_params(self):
...
 
+    def test_clear_param(self):
+        df = self.spark.createDataFrame([(Vectors.dense([1.0]),), (Vectors.dense([2.0]),)], ["a"])
+        maScaler = MaxAbsScaler(inputCol="a", outputCol="scaled")
+        model = maScaler.fit(df)
+        self.assertTrue(model.isSet(model.outputCol))
+        self.assertEqual(model.getOutputCol(), "scaled")
+        model.clear(model.outputCol)
+        self.assertFalse(model.isSet(model.outputCol))
+        self.assertEqual(model.getOutputCol()[:12], 'MaxAbsScaler')
```
Comment (BryanCutler) [inline python/pyspark/ml/tests/test_param.py]: Instead, could you check that `outputCol` is equal to the default value?

Comment (BryanCutler) [reply python/pyspark/ml/tests/test_param.py]: Nvm, I guess there isn't a way to get the default value directly. This is fine then.

**Review by BryanCutler — APPROVED** [review]
LGTM

**Comment by BryanCutler** [comment]
merged to master, thanks @huaxingao !

**Comment by huaxingao** [comment]
Thank you all for the help!

**Pull request closed without merging**